File: js/src/lino.js
/**
 * Links Notation (LiNo) parser and serializer.
 *
 * LiNo represents associative data as nested links. Every link has the form
 *
 *   (index: source target)
 *
 * where `index` (the identity) is optional and the values (`source`, `target`,
 * ...) are themselves references or links. The notation is the surface syntax
 * for the LinksQL substitution model.
 *
 * The grammar implemented here is:
 *
 *   document = { value } ;
 *   value    = link | reference ;
 *   link     = "(" [ value ":" ] { value } ")" ;
 *   reference = number | name | variable | wildcard | string ;
 *   variable = "$" name ;
 *   wildcard = "*" ;
 *
 * The parser produces a small, explicit AST so the rest of the engine never
 * has to re-tokenize text.
 */

/** Error thrown when LiNo input cannot be parsed. */
export class LinoSyntaxError extends Error {
  /**
   * @param {string} message - Human readable description.
   * @param {number} position - Zero-based offset of the offending character.
   */
  constructor(message, position) {
    super(position >= 0 ? `${message} (at position ${position})` : message);
    this.name = 'LinoSyntaxError';
    this.position = position;
  }
}

const WHITESPACE = new Set([' ', '\t', '\n', '\r', '\f', '\v']);
const DELIMITERS = new Set(['(', ')', ':', '"', "'"]);

/** Characters that force a name to be quoted on output. */
const QUOTE_REQUIRED = /[\s():"']/;

/**
 * Classify a bareword (a run of non-delimiter characters) into a reference
 * token. Numbers become numeric references, `$x` becomes a variable, `*`
 * becomes a wildcard, and everything else becomes a name.
 *
 * @param {string} word - The raw bareword.
 * @param {number} pos - Source offset for diagnostics.
 * @returns {object} A reference token.
 */
function classifyWord(word, pos) {
  if (word === '*') {
    return { type: 'ref', refKind: 'wildcard', value: '*', pos };
  }
  if (word[0] === '$') {
    return { type: 'ref', refKind: 'variable', value: word.slice(1), pos };
  }
  if (/^\d+$/.test(word)) {
    return { type: 'ref', refKind: 'number', value: Number(word), pos };
  }
  return { type: 'ref', refKind: 'name', value: word, pos };
}

/**
 * Read a quoted string starting at `start` (which points at the quote).
 *
 * @param {string} input - Full source text.
 * @param {number} start - Index of the opening quote.
 * @returns {{ token: object, next: number }} The string token and next index.
 */
function readString(input, start) {
  const quote = input[start];
  let i = start + 1;
  let value = '';
  while (i < input.length && input[i] !== quote) {
    if (input[i] === '\\' && i + 1 < input.length) {
      value += input[i + 1];
      i += 2;
    } else {
      value += input[i];
      i += 1;
    }
  }
  if (i >= input.length) {
    throw new LinoSyntaxError('Unterminated quoted string', start);
  }
  return {
    token: { type: 'ref', refKind: 'name', value, pos: start, quoted: true },
    next: i + 1,
  };
}

/**
 * Split LiNo source text into a flat list of tokens.
 *
 * @param {string} input - LiNo source text.
 * @returns {object[]} Ordered tokens.
 */
export function tokenize(input) {
  const tokens = [];
  let i = 0;
  while (i < input.length) {
    const ch = input[i];
    if (WHITESPACE.has(ch)) {
      i += 1;
    } else if (ch === '(') {
      tokens.push({ type: 'lparen', pos: i });
      i += 1;
    } else if (ch === ')') {
      tokens.push({ type: 'rparen', pos: i });
      i += 1;
    } else if (ch === ':') {
      tokens.push({ type: 'colon', pos: i });
      i += 1;
    } else if (ch === '"' || ch === "'") {
      const { token, next } = readString(input, i);
      tokens.push(token);
      i = next;
    } else {
      const start = i;
      let word = '';
      while (i < input.length && !WHITESPACE.has(input[i])) {
        if (DELIMITERS.has(input[i])) {
          break;
        }
        word += input[i];
        i += 1;
      }
      tokens.push(classifyWord(word, start));
    }
  }
  return tokens;
}

/** @param {object} state - Parser cursor. */
function peek(state) {
  return state.tokens[state.index];
}

/** @param {object} state - Parser cursor. */
function advance(state) {
  return state.tokens[state.index++];
}

/**
 * Build a reference AST node from a reference token.
 *
 * @param {object} token - A token of type `ref`.
 * @returns {object} Reference node.
 */
function referenceNode(token) {
  return { type: 'ref', kind: token.refKind, value: token.value };
}

/**
 * Parse a single value (a reference or a parenthesized link).
 *
 * @param {object} state - Parser cursor.
 * @returns {object} AST node.
 */
function parseValue(state) {
  const token = peek(state);
  if (!token) {
    throw new LinoSyntaxError('Unexpected end of input', -1);
  }
  if (token.type === 'lparen') {
    return parseLink(state);
  }
  if (token.type === 'ref') {
    advance(state);
    return referenceNode(token);
  }
  throw new LinoSyntaxError(`Unexpected '${token.type}'`, token.pos);
}

/**
 * Parse a link: `(` [ value `:` ] values... `)`.
 *
 * @param {object} state - Parser cursor.
 * @returns {object} Link node.
 */
function parseLink(state) {
  advance(state); // consume '('
  if (peek(state) && peek(state).type === 'rparen') {
    advance(state);
    return { type: 'link', id: null, values: [] };
  }
  const first = parseValue(state);
  let id = null;
  const values = [];
  if (peek(state) && peek(state).type === 'colon') {
    advance(state); // consume ':'
    id = first;
  } else {
    values.push(first);
  }
  while (peek(state) && peek(state).type !== 'rparen') {
    values.push(parseValue(state));
  }
  const closing = peek(state);
  if (!closing || closing.type !== 'rparen') {
    throw new LinoSyntaxError('Expected )', closing ? closing.pos : -1);
  }
  advance(state); // consume ')'
  return { type: 'link', id, values };
}

/**
 * Parse LiNo source text into an array of top-level AST nodes.
 *
 * @param {string} input - LiNo source text.
 * @returns {object[]} Top-level nodes (links and references).
 */
export function parse(input) {
  if (typeof input !== 'string') {
    throw new LinoSyntaxError('LiNo input must be a string', -1);
  }
  const state = { tokens: tokenize(input), index: 0 };
  const values = [];
  while (state.index < state.tokens.length) {
    values.push(parseValue(state));
  }
  return values;
}

/**
 * Quote a name for output if it contains characters that would otherwise be
 * interpreted as structure.
 *
 * @param {string} name - Raw name.
 * @returns {string} Possibly quoted name.
 */
function quoteName(name) {
  if (name === '' || QUOTE_REQUIRED.test(name)) {
    return `"${name.replace(/\\/g, '\\\\').replace(/"/g, '\\"')}"`;
  }
  return name;
}

/**
 * Serialize a reference node to LiNo text.
 *
 * @param {object} node - Reference node.
 * @returns {string} LiNo text.
 */
function serializeRef(node) {
  if (node.kind === 'variable') {
    return `$${node.value}`;
  }
  if (node.kind === 'wildcard') {
    return '*';
  }
  if (node.kind === 'number') {
    return String(node.value);
  }
  return quoteName(String(node.value));
}

/**
 * Serialize an AST node (reference or link) back to LiNo text.
 *
 * @param {object} node - AST node.
 * @returns {string} LiNo text.
 */
export function serialize(node) {
  if (node.type === 'ref') {
    return serializeRef(node);
  }
  const parts = node.values.map(serialize);
  const body = parts.join(' ');
  if (node.id !== null && node.id !== undefined) {
    const id = serialize(node.id);
    return body ? `(${id}: ${body})` : `(${id}:)`;
  }
  return `(${body})`;
}

/**
 * Serialize an array of top-level nodes to LiNo text, one node per joiner.
 *
 * @param {object[]} nodes - Top-level nodes.
 * @param {string} [joiner] - Separator between nodes (default newline).
 * @returns {string} LiNo text.
 */
export function serializeAll(nodes, joiner = '\n') {
  return nodes.map(serialize).join(joiner);
}

File: rust/src/lino.rs
//! Links Notation (LiNo) parser and serializer.
//!
//! LiNo represents associative data as nested links. Every link has the form
//!
//! ```text
//! (index: source target)
//! ```
//!
//! where `index` (the identity) is optional and the values (`source`, `target`,
//! ...) are themselves references or links. The notation is the surface syntax
//! for the LinksQL substitution model.
//!
//! The grammar implemented here is:
//!
//! ```text
//! document  = { value } ;
//! value     = link | reference ;
//! link      = "(" [ value ":" ] { value } ")" ;
//! reference = number | name | variable | wildcard | string ;
//! variable  = "$" name ;
//! wildcard  = "*" ;
//! ```
//!
//! The parser produces a small, explicit AST so the rest of the engine never has
//! to re-tokenize text.

use crate::{Error, Result};

/// A reference: the leaf of the AST.
///
/// Mirrors the JavaScript reference node `{ type: "ref", kind, value }`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Ref {
    /// A numeric reference (an integer index).
    Number(i64),
    /// A named reference (a human-readable label).
    Name(String),
    /// A variable reference, written `$name`; carries the name without the `$`.
    Variable(String),
    /// The wildcard `*`.
    Wildcard,
}

/// An AST node: either a reference leaf or a (possibly nested) link.
///
/// Mirrors the JavaScript node union of reference nodes and link nodes.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Node {
    /// A reference leaf.
    Ref(Ref),
    /// A link. `id` is the optional identity (itself a node, so it may be a
    /// reference or a nested link); `values` is the ordered value list.
    Link {
        /// Optional identity node.
        id: Option<Box<Self>>,
        /// Ordered values.
        values: Vec<Self>,
    },
}

/// A token produced by [`tokenize`].
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Token {
    /// `(`
    LParen,
    /// `)`
    RParen,
    /// `:`
    Colon,
    /// A reference token (bareword or quoted string).
    Ref(Ref),
}

/// Characters treated as whitespace (token separators, otherwise insignificant):
/// space, tab, newline, carriage return, form feed (`\x0c`), vertical tab
/// (`\x0b`).
const fn is_whitespace(ch: char) -> bool {
    matches!(ch, ' ' | '\t' | '\n' | '\r' | '\u{0c}' | '\u{0b}')
}

/// Characters that terminate a bareword.
const fn is_delimiter(ch: char) -> bool {
    matches!(ch, '(' | ')' | ':' | '"' | '\'')
}

/// Characters that force a name to be quoted on output: whitespace, `(`, `)`,
/// `:`, `"`, `'`.
fn quote_required(name: &str) -> bool {
    name.chars()
        .any(|ch| is_whitespace(ch) || matches!(ch, '(' | ')' | ':' | '"' | '\''))
}

/// Classify a bareword (a run of non-delimiter characters) into a reference
/// token. Numbers become numeric references, `$x` becomes a variable, `*`
/// becomes a wildcard, and everything else becomes a name.
fn classify_word(word: &str) -> Ref {
    if word == "*" {
        return Ref::Wildcard;
    }
    if let Some(rest) = word.strip_prefix('$') {
        return Ref::Variable(rest.to_string());
    }
    if !word.is_empty() && word.bytes().all(|b| b.is_ascii_digit()) {
        // `^[0-9]+$`: parse as an integer index.
        if let Ok(value) = word.parse::<i64>() {
            return Ref::Number(value);
        }
    }
    Ref::Name(word.to_string())
}

/// Split LiNo source text into a flat list of tokens.
///
/// # Examples
///
/// ```
/// use linksql::{tokenize, Token, Ref};
/// let kinds = tokenize("(1: 1 1)").unwrap();
/// assert!(matches!(kinds[0], Token::LParen));
/// assert!(matches!(kinds[2], Token::Colon));
/// assert!(matches!(kinds[5], Token::RParen));
/// ```
pub fn tokenize(input: &str) -> Result<Vec<Token>> {
    // Work over Unicode scalar values so positions match the JavaScript engine,
    // which indexes by UTF-16 code unit for the BMP characters used by LiNo.
    let chars: Vec<char> = input.chars().collect();
    let mut tokens = Vec::new();
    let mut i = 0usize;
    while i < chars.len() {
        let ch = chars[i];
        if is_whitespace(ch) {
            i += 1;
        } else if ch == '(' {
            tokens.push(Token::LParen);
            i += 1;
        } else if ch == ')' {
            tokens.push(Token::RParen);
            i += 1;
        } else if ch == ':' {
            tokens.push(Token::Colon);
            i += 1;
        } else if ch == '"' || ch == '\'' {
            let (value, next) = read_string(&chars, i)?;
            // A quoted token is ALWAYS a name, even if it looks numeric/`$x`/`*`.
            tokens.push(Token::Ref(Ref::Name(value)));
            i = next;
        } else {
            let start = i;
            let mut word = String::new();
            while i < chars.len() && !is_whitespace(chars[i]) {
                if is_delimiter(chars[i]) {
                    break;
                }
                word.push(chars[i]);
                i += 1;
            }
            // `start` is unused beyond clarity, but kept to mirror the source
            // structure of the JavaScript tokenizer.
            let _ = start;
            tokens.push(Token::Ref(classify_word(&word)));
        }
    }
    Ok(tokens)
}

/// Read a quoted string starting at `start` (which points at the quote).
///
/// Returns the unescaped value and the index just past the closing quote.
fn read_string(chars: &[char], start: usize) -> Result<(String, usize)> {
    let quote = chars[start];
    let mut i = start + 1;
    let mut value = String::new();
    while i < chars.len() && chars[i] != quote {
        if chars[i] == '\\' && i + 1 < chars.len() {
            // Backslash escapes the next character (`\"`→`"`, `\\`→`\`).
            value.push(chars[i + 1]);
            i += 2;
        } else {
            value.push(chars[i]);
            i += 1;
        }
    }
    if i >= chars.len() {
        // Positions are small, but convert fallibly to satisfy the lint and stay
        // safe; an impossibly large position degrades to "no position" (-1).
        let position = i64::try_from(start).unwrap_or(-1);
        return Err(Error::lino_syntax("Unterminated quoted string", position));
    }
    Ok((value, i + 1))
}

/// Internal parser cursor over a token slice.
struct Parser<'a> {
    tokens: &'a [Token],
    index: usize,
}

impl<'a> Parser<'a> {
    const fn new(tokens: &'a [Token]) -> Self {
        Self { tokens, index: 0 }
    }

    fn peek(&self) -> Option<&Token> {
        self.tokens.get(self.index)
    }

    fn advance(&mut self) -> Option<&Token> {
        let token = self.tokens.get(self.index);
        if token.is_some() {
            self.index += 1;
        }
        token
    }

    /// Parse a single value (a reference or a parenthesized link).
    fn parse_value(&mut self) -> Result<Node> {
        match self.peek() {
            None => Err(Error::lino_syntax("Unexpected end of input", -1)),
            Some(Token::LParen) => self.parse_link(),
            Some(Token::Ref(r)) => {
                let node = Node::Ref(r.clone());
                self.advance();
                Ok(node)
            }
            // `)` or `:` cannot start a value. The JavaScript parser names the
            // token type; we reproduce the same diagnostic shape.
            Some(Token::RParen) => Err(Error::lino_syntax("Unexpected 'rparen'", -1)),
            Some(Token::Colon) => Err(Error::lino_syntax("Unexpected 'colon'", -1)),
        }
    }

    /// Parse a link: `(` `[` value `:` `]` values... `)`.
    fn parse_link(&mut self) -> Result<Node> {
        self.advance(); // consume '('
        if matches!(self.peek(), Some(Token::RParen)) {
            self.advance();
            return Ok(Node::Link {
                id: None,
                values: Vec::new(),
            });
        }
        let first = self.parse_value()?;
        let mut id = None;
        let mut values = Vec::new();
        if matches!(self.peek(), Some(Token::Colon)) {
            self.advance(); // consume ':'
            id = Some(Box::new(first));
        } else {
            values.push(first);
        }
        while let Some(token) = self.peek() {
            if matches!(token, Token::RParen) {
                break;
            }
            values.push(self.parse_value()?);
        }
        if !matches!(self.peek(), Some(Token::RParen)) {
            return Err(Error::lino_syntax("Expected )", -1));
        }
        self.advance(); // consume ')'
        Ok(Node::Link { id, values })
    }
}

/// Parse LiNo source text into an array of top-level AST nodes.
///
/// # Examples
///
/// ```
/// use linksql::{parse, Node, Ref};
/// assert_eq!(parse("1").unwrap(), vec![Node::Ref(Ref::Number(1))]);
/// ```
pub fn parse(input: &str) -> Result<Vec<Node>> {
    let tokens = tokenize(input)?;
    let mut parser = Parser::new(&tokens);
    let mut values = Vec::new();
    while parser.index < parser.tokens.len() {
        values.push(parser.parse_value()?);
    }
    Ok(values)
}

/// Quote a name for output if it contains characters that would otherwise be
/// interpreted as structure, escaping `\` and `"`.
fn quote_name(name: &str) -> String {
    if name.is_empty() || quote_required(name) {
        let escaped = name.replace('\\', "\\\\").replace('"', "\\\"");
        format!("\"{escaped}\"")
    } else {
        name.to_string()
    }
}

/// Serialize a reference node to LiNo text.
fn serialize_ref(r: &Ref) -> String {
    match r {
        Ref::Variable(name) => format!("${name}"),
        Ref::Wildcard => "*".to_string(),
        Ref::Number(value) => value.to_string(),
        Ref::Name(name) => quote_name(name),
    }
}

/// Serialize an AST node (reference or link) back to LiNo text.
///
/// # Examples
///
/// ```
/// use linksql::{parse, serialize};
/// let node = &parse("(name: \"hello world\")").unwrap()[0];
/// assert_eq!(serialize(node), "(name: \"hello world\")");
/// ```
#[must_use]
pub fn serialize(node: &Node) -> String {
    match node {
        Node::Ref(r) => serialize_ref(r),
        Node::Link { id, values } => {
            let body = values.iter().map(serialize).collect::<Vec<_>>().join(" ");
            id.as_ref().map_or_else(
                || format!("({body})"),
                |id_node| {
                    let id = serialize(id_node);
                    if body.is_empty() {
                        format!("({id}:)")
                    } else {
                        format!("({id}: {body})")
                    }
                },
            )
        }
    }
}

/// Serialize an array of top-level nodes to LiNo text, one node per joiner.
///
/// The default joiner used by the library elsewhere is a newline; tests pass a
/// space to compare against the canonical round-trip samples.
#[must_use]
pub fn serialize_all(nodes: &[Node], joiner: &str) -> String {
    nodes.iter().map(serialize).collect::<Vec<_>>().join(joiner)
}
